File: assets/Scripts/player/controllers/PlayerEventsController.js
"use strict";

const PlayerEvents = require("../PlayerEvents.js");

class PlayerEventsController {
    constructor() {
        this.ctx = null;
        this.entity = null;
        this.impl = null;
    }

    bind(ctx, entity) {
        this.ctx = ctx;
        this.entity = entity;
        return this;
    }

    onStart() {
        this.impl = new PlayerEvents(this.entity);

        if (this.impl && typeof this.impl.emit === "function") {
            const uuid = (this.entity && typeof this.entity.uuidString === "function")
                ? this.entity.uuidString()
                : (this.entity && typeof this.entity.uuid === "function")
                    ? this.entity.uuid()
                    : this.entity && this.entity.uuid;
            this.impl.emit("player.spawn", {
                uuid,
                bodyId: this.entity.bodyId
            });
        }

        const ctx = this.entity.ctx;
        if (ctx && typeof ctx.state === "function") {
            ctx.state().set("player", {
                alive: true,
                uuid: (this.entity && typeof this.entity.uuidString === "function")
                    ? this.entity.uuidString()
                    : (this.entity && typeof this.entity.uuid === "function")
                        ? this.entity.uuid()
                        : this.entity && this.entity.uuid,
                surfaceId: this.entity.surfaceId,
                bodyId: this.entity.bodyId
            });
        }
    }

    onUpdate(dt) {
        if (this.impl && typeof this.impl.tick === "function") {
            this.impl.tick(this.entity.frame);
        }
    }

    onStop() {
        if (this.impl && typeof this.impl.destroy === "function") this.impl.destroy();
        this.impl = null;

        const ctx = this.entity.ctx;
        if (ctx && typeof ctx.state === "function") ctx.state().remove("player");
    }
}

module.exports = {PlayerEventsController};

File: assets/Scripts/player/controllers/PlayerEventsController.lua
local M = {}
local luaRuntime = require("@builtin/lua_runtime")
local LuaClass = luaRuntime.LuaClass
local LuaConstruct = luaRuntime.LuaConstruct
PlayerEvents = require("../PlayerEvents.lua")
PlayerEventsController = LuaClass()
PlayerEventsController.name = "PlayerEventsController"
function PlayerEventsController.prototype.lua_constructor(self)
    self.ctx = nil
    self.entity = nil
    self.impl = nil
end
function PlayerEventsController.prototype.bind(self, ctx, entity)
    self.ctx = ctx
    self.entity = entity
    return self
end
function PlayerEventsController.prototype.onStart(self)
    self.impl = LuaConstruct(PlayerEvents, self.entity)
    if self.impl and KTypeOf(self.impl.emit) == "function" then
        local lua_temp_1
        if self.entity and KTypeOf(self.entity.uuidString) == "function" then
            lua_temp_1 = self.entity:uuidString()
        else
            local lua_temp_0
            if self.entity and KTypeOf(self.entity.uuid) == "function" then
                lua_temp_0 = self.entity:uuid()
            else
                lua_temp_0 = self.entity and self.entity.uuid
            end
            lua_temp_1 = lua_temp_0
        end
        local uuid = lua_temp_1
        self.impl:emit("player.spawn", {uuid = uuid, bodyId = self.entity.bodyId})
    end
    local ctx = self.entity.ctx
    if ctx and KTypeOf(ctx.state) == "function" then
        local lua_self_4 = ctx:state()
        local lua_self_4_set_5 = lua_self_4.set
        local lua_temp_3
        if self.entity and KTypeOf(self.entity.uuidString) == "function" then
            lua_temp_3 = self.entity:uuidString()
        else
            local lua_temp_2
            if self.entity and KTypeOf(self.entity.uuid) == "function" then
                lua_temp_2 = self.entity:uuid()
            else
                lua_temp_2 = self.entity and self.entity.uuid
            end
            lua_temp_3 = lua_temp_2
        end
        lua_self_4_set_5(lua_self_4, "player", {alive = true, uuid = lua_temp_3, surfaceId = self.entity.surfaceId, bodyId = self.entity.bodyId})
    end
end
function PlayerEventsController.prototype.onUpdate(self, dt)
    if self.impl and KTypeOf(self.impl.tick) == "function" then
        self.impl:tick(self.entity.frame)
    end
end
function PlayerEventsController.prototype.onStop(self)
    if self.impl and KTypeOf(self.impl.destroy) == "function" then
        self.impl:destroy()
    end
    self.impl = nil
    local ctx = self.entity.ctx
    if ctx and KTypeOf(ctx.state) == "function" then
        ctx:state():remove("player")
    end
end
M = {PlayerEventsController = PlayerEventsController}

return M
